### PR TITLE
snap: reenable i386, arm64, armhf architecture publishing

### PR DESCRIPTION
**Pull request by 5chdn**

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -116,9 +131,29 @@ publish-snap:
...
+  <<:                              *publish-snap
+  variables:
+    BUILD_ARCH:                  arm64
+  dependencies:
+    - build-snap-arm64
+
+publish-snap-armhf:
+  <<:                              *publish-snap
+  variables:
+    BUILD_ARCH:                  armhf
```
Comment (TriplEight) [inline .gitlab-ci.yml]: 1 more Tab plz ^^^ :)

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -67,6 +67,21 @@ build-linux:
...
+
+build-snap-arm64:
+  <<:                              *build-linux
+  variables:
+    CARGO_TARGET:                  aarch64-unknown-linux-gnu
+
+build-snap-armhf:
+  <<:                              *build-linux
+  variables:
+    CARGO_TARGET:                  armv7-unknown-linux-gnueabihf
```
Comment (TriplEight) [inline .gitlab-ci.yml]: If these changes will be merged, would be smart to preinstall these targets in the new Dockerfile. And push it to production right away.

**Comment by TriplEight** [comment]
not sure though, if all these builds can be done in the same Docker image, have to check first.

**Comment by General-Beck** [comment]
@elopio any ideas? https://gitlab.parity.io/parity/parity-ethereum/-/jobs/124182
other tasks with other architectures were built successfully

**Comment by 5chdn** [comment]
@General-Beck this also fails on nightly, so probably unrelated

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -4,8 +4,7 @@ stages:
   - publish
   - optional
 
-image:                             parity/rust:gitlab-ci
-
+image:                             parity/rust-parity-ethereum-build:xenial
```
Comment (soc1c) [inline .gitlab-ci.yml]: what's the logic behind changing this now?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -35,14 +34,6 @@ variables:
     paths:
       - artifacts/
 
-.determine_version:                &determine_version
-  - VERSION="$(sed -r -n '1,/^version/s/^version = "([^"]+)".*$/\1/p' Cargo.toml)"
```
Comment (soc1c) [inline .gitlab-ci.yml]: Why is this removed?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -108,24 +117,43 @@ publish-docker:
...
   image:                           snapcore/snapcraft
   variables:
     BUILD_ARCH:                    amd64
   cache: {}
-  before_script:                   *determine_version
   dependencies:
     - build-linux
   tags:
-    - rust-stable
+    - linux-docker
```
Comment (soc1c) [inline .gitlab-ci.yml]: Why `linux-docker`?

File: scripts/gitlab/build-unix.sh [hunk 6]
```diff
@@ -43,11 +48,6 @@ fi
...
 for binary in $(ls)
 do
-  rhash --sha256 $binary -o $binary.sha256
-  if [ "${CARGO_TARGET}" = "armv7-linux-androideabi" ]
-  then
-    echo "> ${binary} cannot be hashed with cross-compiled binary"
-  else
-    ./parity tools hash $binary > $binary.sha3
-  fi
+  rhash --sha256 $binary -o $binary.sha256 #do we still need this hash (SHA2)?
```
Comment (soc1c) [inline scripts/gitlab/build-unix.sh]: SHA256 is used for Vanity Service and Homebrew (Releases) And One-line Installer

File: scripts/gitlab/publish-snap.sh [hunk 7]
```diff
@@ -8,7 +8,10 @@ set -u # treat unset variables as error
 # gsub(/ /, "", $2) deletes whitespaces
 TRACK=`awk -F '=' '/^track/ {gsub(/"/, "", $2); gsub(/ /, "", $2); print $2}' ./util/version/Cargo.toml`
 echo Track is: $TRACK
-
+# prepare variables
+VERSION=v"$(sed -r -n '1,/^version/s/^version = "([^"]+)".*$/\1/p' Cargo.toml)"
+SNAP_PACKAGE="parity_"$VERSION"_"$BUILD_ARCH".snap"
+CARGO_TARGET="$(ls artifacts)"
```
Comment (soc1c) [inline scripts/gitlab/publish-snap.sh]: What? `ls artifacts` gives a reliable `CARGO_TARGET`?

File: scripts/snap/snapcraft.template.yaml [hunk 8]
```diff
@@ -50,4 +50,4 @@ parts:
       cp -v ethkey $SNAPCRAFT_PART_INSTALL/usr/bin/ethkey
       cp -v ethstore $SNAPCRAFT_PART_INSTALL/usr/bin/ethstore
       cp -v whisper $SNAPCRAFT_PART_INSTALL/usr/bin/whisper
-    stage-packages: [libc6, libudev1, libstdc++6, cmake, libdb5.3]
```
Comment (soc1c) [inline scripts/snap/snapcraft.template.yaml]: we added libc6 because it was failling without. It's working now?

File: .gitlab-ci.yml (hunk 3 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: this docker image has been tested,
it's the preparation for using cache in docker volumes

File: .gitlab-ci.yml (hunk 4 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: was used only in `publish-snap.sh` so moved there.
Is there a reason for it to stay here?

File: scripts/gitlab/build-unix.sh (hunk 6 above)
Comment (TriplEight) [reply scripts/gitlab/build-unix.sh]: :+1:

File: scripts/snap/snapcraft.template.yaml (hunk 8 above)
Comment (TriplEight) [reply scripts/snap/snapcraft.template.yaml]: and it was the reason it failed here: https://github.com/paritytech/devops/issues/307

File: scripts/gitlab/build-unix.sh [hunk 9]
```diff
@@ -43,11 +48,6 @@ fi
...
+  rhash --sha3-256 $binary -o $binary.sha3
```
Comment (gabreal) [inline scripts/gitlab/build-unix.sh]: hashes are different, is that okay?
```
% target/release/parity tools hash target/release/parity
58b49924c583b45871cac4077da8f99820a684df8cca9a81c742080e991f671d
% rhash --sha3-256 target/release/parity
95fcba97a2059bbae16f997c3ea7b6c6c0f9ef5990313f55eb8536a1b3bef9e3  target/release/parity
% rhash --sha256 target/release/parity                
19c7ac31b99f5a45e6a4236367df18467321a3fc9f492c19f80cede098d47abf  target/release/parity
```

File: scripts/gitlab/build-unix.sh [hunk 10]
```diff
@@ -11,6 +11,11 @@ echo "CC:               " $CC
 echo "CXX:              " $CXX
 #strip ON
 export RUSTFLAGS=" -C link-arg=-s"
+# Linker for crosscomile
+echo "[target.$CARGO_TARGET]" >> .cargo/config
+echo "linker= \"$CC\"" >> .cargo/config
```
Comment (gabreal) [inline scripts/gitlab/build-unix.sh]: config file will grow once caching is back enabled.

File: .gitlab-ci.yml (hunk 3 above)
Comment (gabreal) [inline .gitlab-ci.yml]: would prefer a separate pr for that, just in case one has to revert.

File: scripts/gitlab/publish-snap.sh (hunk 7 above)
Comment (General-Beck) [reply scripts/gitlab/publish-snap.sh]: yes, from artifacts dir

File: .gitlab-ci.yml (hunk 5 above)
Comment (General-Beck) [reply .gitlab-ci.yml]: clean up the tags

File: .gitlab-ci.yml (hunk 4 above)
Comment (General-Beck) [reply .gitlab-ci.yml]: used only for publish snap

File: .gitlab-ci.yml (hunk 3 above)
Comment (General-Beck) [reply .gitlab-ci.yml]: new CI images

File: scripts/snap/snapcraft.template.yaml (hunk 8 above)
Comment (General-Beck) [reply scripts/snap/snapcraft.template.yaml]: yes

File: scripts/gitlab/build-unix.sh (hunk 9 above)
Comment (General-Beck) [reply scripts/gitlab/build-unix.sh]: we can't use parity to calculate the hash for other architectures

File: scripts/gitlab/build-unix.sh (hunk 10 above)
Comment (General-Beck) [reply scripts/gitlab/build-unix.sh]: we will not use caching for this task. it is not necessary

Comment (General-Beck) [reply scripts/gitlab/build-unix.sh]: fixed

**Comment by soc1c** [comment]
Needs a 2nd review.

**Comment by TriplEight** [comment]
android parity-clib compiles now: https://gitlab.parity.io/parity/parity-ethereum/pipelines/31445

**Comment by TriplEight** [comment]
The latest pipeline is all green: https://gitlab.parity.io/parity/parity-ethereum/pipelines/31461
